hpx/lcos/dist_object.hpp

```diff
@@ -0,0 +1,566 @@
+        HPX_PP_CAT(__dist_object_part_fetch_action_, type));                   \
+    typedef ::hpx::components::component<                                      \
+        hpx::lcos::server::dist_object_part<type>>                             \
+        HPX_PP_CAT(__dist_object_part_, type);                                 \
+    HPX_REGISTER_COMPONENT(HPX_PP_CAT(__dist_object_part_, type))              \
+    /**/
+
+namespace hpx { namespace lcos {
+    enum class construction_type
+    {
+        Meta_Object,
+        All_to_All
+    };
+}}
+
+namespace hpx { namespace lcos {
+    class meta_object_server
+      : public hpx::components::locking_hook<
+            hpx::components::component_base<meta_object_server>>
+    {
+    public:
+        meta_object_server()
+          : b(hpx::find_all_localities().size())
+        {
+            servers.resize(hpx::find_all_localities().size());
+        }
+
+        std::vector<hpx::id_type> get_server_list()
+        {
+            if (servers.size() == hpx::find_all_localities().size())
+            {
+                num_sent++;
+                return servers;
+            }
+            std::vector<hpx::id_type> empty;
+            empty.resize(0);
+            return empty;
+        }
+
+        std::vector<hpx::id_type> registration(
+            std::size_t source_loc, hpx::id_type id)
+        {
+            {
+                // TODO: duplicated lock: the namespace has lock guard already
+                std::lock_guard<hpx::lcos::local::spinlock> l(lk);
+                servers[source_loc] = id;
+            }
+            b.wait();
+            return servers;
+        }
+
+        HPX_DEFINE_COMPONENT_ACTION(meta_object_server, get_server_list);
+        HPX_DEFINE_COMPONENT_ACTION(meta_object_server, registration);
+
+    private:
+        int num_sent;
+        hpx::lcos::local::barrier b;
+        hpx::lcos::local::spinlock lk;
+        std::vector<hpx::id_type> servers;
+    };
+}}
+typedef hpx::lcos::meta_object_server::get_server_list_action get_list_action;
+HPX_REGISTER_ACTION_DECLARATION(get_list_action, get_server_list_mo_action);
+typedef hpx::lcos::meta_object_server::registration_action
+    register_with_meta_action;
+HPX_REGISTER_ACTION_DECLARATION(register_with_meta_action, register_mo_action);
+
+// Meta_object front end, decides whether it is the root locality, and thus
+// whether to register with the root locality's meta object only or to register
+// itself as the root locality's meta object as well
+namespace hpx { namespace lcos {
+    class meta_object
+      : hpx::components::client_base<meta_object, meta_object_server>
+    {
+    public:
+        typedef hpx::components::client_base<meta_object, meta_object_server>
+            base_type;
+
+        meta_object(std::string basename)
+          : base_type(hpx::local_new<meta_object_server>())
+        {
+            if (hpx::get_locality_id() == 0)
```

Comment:
Thanks for your suggestion. Actually, it can be any locality and being in an arbitrary locality adds flexibility. See here: https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/hpx/lcos/distributed_object.hpp#L219

Now, we sets the root to the first locality of the given sub-locality list. For example, If all localities (default option) are given to construct `distributed_object`, then the root will be locality 0. Say, if you provide localities `1, 3, 5`, then the locality `1` will be the root for the meta_object. Let me know if this fits for you.